File: src/main/java/algo/slidingWindow/FindAllAnagramsInAString_438.java
package algo.slidingWindow;

import java.util.ArrayList;
import java.util.List;

public class FindAllAnagramsInAString_438 {

    public static List<Integer> findAnagrams(String s, String p) {
        var result = new ArrayList<Integer>();
        var window = p.length();

        if (s.length() < window) {
            return result;
        }

        var freq = new int[26];
        for (var c : p.toCharArray()) {
            freq[c - 'a']++;
        }

        var count = new int[26];
        for (var i = 0; i < window; i++) {
            count[s.charAt(i) - 'a']++;
        }

        for (var i = 0; i <= s.length() - window; i++) {
            if (i > 0) {
                count[s.charAt(i - 1) - 'a']--;
                count[s.charAt(i + window - 1) - 'a']++;
            }
            if (match(count, freq)) {
                result.add(i);
                continue;
            }
        }

        return result;
    }

    private static boolean match(int[] count, int[] freq) {
        for (var i = 0; i < count.length; i++) {
            if (count[i] != freq[i]) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println(findAnagrams("cbaebabacd", "abc"));
        System.out.println(findAnagrams("abab", "ab"));
    }
}

File: src/main/kotlin/algo/slidingWindow/FindAllAnagramsInAString_438.kt
package algo.slidingWindow

class KtFindAllAnagramsInAString_438 {
    fun findAnagrams(s: String, p: String): List<Int> {
        val starts: MutableList<Int> = mutableListOf()
        if (p.length > s.length) {
            return starts
        }

        val base = IntArray(26)
        p.forEach { base[it - 'a']++ }

        val count = IntArray(26)
        s.indices.forEach { i ->
            val ch = s[i]
            if (i < p.length) {
                count[ch - 'a']++
            } else {
                if (isAnagram(base, count)) {
                    starts.add(i - p.length)
                }
                count[ch - 'a']++
                count[s[i - p.length] - 'a']--
            }
        }
        if (isAnagram(base, count)) {
            starts.add(s.length - p.length)
        }
        return starts
    }

    private fun isAnagram(a: IntArray, b: IntArray): Boolean {
        for (j in a.indices) {
            if (a[j] != b[j]) {
                return false
            }
        }
        return true

    }
}

fun main() {
    val anagram = KtFindAllAnagramsInAString_438()
    println(anagram.findAnagrams("cbaebabacd", "abc"))
    println(anagram.findAnagrams("abab", "ab"))
}